### ISSUE TITLE
mapreduce: sharding issues

### ISSUE THREAD

**Issue by jacobg** [potatolondon/djangae#506]
There’s a couple issues with your split_input implementation:
(1) I’m not finding balanced distribution across shards. Some shards are even totally unused. It's possible this may occur for older applications that include older model entities that used the old sequential id strategy rather than the newer scattered one. But I'm not sure if there could be other reasons as well.
(2) It requires a -pk datastore index for the model, which official mapreduce AbstractDatastoreInputReader does not.

It looks like the official mapreduce library implements its shard splitter by ordering by a model property called **scatter** and sampling datastore keys, which it then uses to generate a key range for each shard. Using a Google Search, I discovered that this **scatter** property is used to get random distributed samples in the keyspace. The lowest key range has no starting key, and the highest key range has no ending key. Perhaps we should do the same: https://github.com/GoogleCloudPlatform/appengine-mapreduce/blob/71ed723e494dbcbe3ea77220612b938abeb452a5/java/src/main/java/com/google/appengine/tools/mapreduce/inputs/DatastoreShardStrategy.java#L58

Maybe a solution would be to use context manager when querying for samples to notify the SelectCommand to order by **scatter**, as long as there are no inequality filters.

**Comment by Kazade** [comment]
I'd quite like to expose ordering by `__scatter__` in Djangae, it should be trivial to add. There is some work ongoing to provide a better API for using MR under Djangae but I don't know the status, I'll find out!

**Comment by jacobg** [comment]
Thanks. I'm still reading through the mapreduce code to try to understand what `__scatter__` is. It seems to me that it refers to a shadow index that gets created along with every other index that contains random samplings of keys for entities in its parent index. In the mapreduce code, it tried to oversample the # of keys in comparison to shards, and then reduce them down to the shard count. But there are comments that sometimes it's possible not to get any of these scattered keys back, I guess due to their randomness. In that case, it gives the entire key range to one shard. Or it could get less than the # of shards, in which case shards get left unused.

**Comment by jacobg** [comment]
Also, if the queryset contains inequality filter, then the official mapreduce library will instead naively split the range using bisections based on fixed constants for the min/max key values. There's a lot of details to it due to namespaces and how the namespace__key gets converted into a number lexicographically. The code is in mapreduce.namespace_range. If djangae is not going to support namespaces, you can probably just shard the range [0, 2^52 + 2^51]. The max id can be seen in google.appengine.datastore.datastore_stub_util:
```
_MAX_SEQUENTIAL_BIT = 52
_MAX_SEQUENTIAL_COUNTER = (1 << _MAX_SEQUENTIAL_BIT) - 1
_MAX_SEQUENTIAL_ID = _MAX_SEQUENTIAL_COUNTER
_MAX_SCATTERED_COUNTER = (1 << (_MAX_SEQUENTIAL_BIT - 1)) - 1
_MAX_SCATTERED_ID = _MAX_SEQUENTIAL_ID + 1 + _MAX_SCATTERED_COUNTER
```

**Comment by chargrizzle** [comment]
Native MapReduce uses a clustering strategy in combination with __scatter__ to ensure a roughly even shard distribution, implementation is in the pipeline and should be completed soon, unfortunately we were waiting on a PR request to be merged into the Pipeline library for this to happen.

**Comment by jacobg** [comment]
Thanks. For inequality filters, will you also fall back to splitting key ranges using lexicographic bisections of the entire global key range?

**Comment by chargrizzle** [comment]
https://github.com/potatolondon/djangae/pull/536
